### PR TITLE
LSM

### PR DESCRIPTION
**Pull request by denkle**

### CODE REVIEW

File: sparsecoding/inference.py [hunk 1]
```diff
@@ -269,4 +269,112 @@ def infer(self, data, dictionary):
...
+    
+    
+    
+class LSM(InferenceMethod):
+    """
+    Infer coefficients for each image in data using elements dictionary.
+    Method implemented according to "Group Sparse Coding with a Laplacian Scale Mixture Prior" (P. J. Garrigues & B. A. Olshausen, 2010)
+    """    
+
+    def __init__(self, n_iter = 100, n_iter_LSM = 6, BETA=0.01, ALPHA=80.0, sigma=0.005, sparseThreshold = 10**-2, solver = None):
```
Comment (cshewmake2) [inline sparsecoding/inference.py:281]: make all variable names lower case and snake case

camel -> snake
sparseThreshold -> sparse_threshold

File: sparsecoding/inference.py [hunk 2]
```diff
@@ -269,4 +269,112 @@ def infer(self, data, dictionary):
...
+            threshold used to discard smallest coefficients in the final solution
+            SM parameter used to compute the loss function
+            
+        solver : default=None
+        '''
+        super().__init__(solver)
+        self.n_iter = n_iter
+        self.n_iter_LSM = n_iter_LSM
+        
+        self.BETA = BETA
```
Comment (cshewmake2) [inline sparsecoding/inference.py:311]: Same here

File: sparsecoding/inference.py [hunk 3]
```diff
@@ -269,4 +269,112 @@ def infer(self, data, dictionary):
...
+        self.ALPHA = ALPHA
+        self.sigma = sigma
+        self.sparseThreshold = sparseThreshold
+
+    def LSMLoss(self, Phi, x, s, Lambda, sigma):
```
Comment (cshewmake2) [inline sparsecoding/inference.py:316]: All functions should have names which are lower case / snake case.

LSMLoss -> lsm_loss

File: sparsecoding/inference.py [hunk 4]
```diff
@@ -269,4 +269,112 @@ def infer(self, data, dictionary):
...
+        n_features, n_basis = dictionary.shape
+        device = dictionary.device
+
+        # initialize
+        #u = torch.zeros((batch_size, n_basis)).to(device)
+        coefficients=torch.zeros(batch_size, n_basis, requires_grad=True).to(device)
+
+        for i in range(0,self.n_iter_LSM):
+        
+            Lambda = (self.ALPHA + 1)/(self.BETA + torch.abs(coefficients))
```
Comment (cshewmake2) [inline sparsecoding/inference.py:346]: Same comment on variable names here. 

I'm okay with python variable names having upper case parts, e.g. I and Phi...
`lambda` is a special function in python, so maybe call it something else or keep the capital letter at the start

File: sparsecoding/inference.py [hunk 5]
```diff
@@ -269,4 +269,112 @@ def infer(self, data, dictionary):
...
+        
+            coefficients=torch.zeros(batch_size, n_basis, requires_grad=True)
+            optimizer = torch.optim.Adam([coefficients])
+        
+            for t in range(0,self.n_iter):
+                
+                loss = self.LSMLoss(Phi=dictionary, x=data, s=coefficients, Lambda=Lambda, sigma=self.sigma) # compute LSM loss
```
Comment (cshewmake2) [inline sparsecoding/inference.py:353]: Move comment to above the line instead of beside

File: sparsecoding/inference.py [hunk 6]
```diff
@@ -269,4 +269,112 @@ def infer(self, data, dictionary):
...
+                optimizer.zero_grad()
+        
+                # Backward pass: compute gradient of the loss with respect to model parameters
+                loss.backward(torch.ones((batch_size,1)),retain_graph=True)
+        
+                # Calling the step function on an Optimizer makes an update to its parameters
+                optimizer.step()
+        
+        
+        coefficients.data[torch.abs(coefficients.data)<self.sparseThreshold] = 0 # Sparsify the final solution by discarding the small coefficients
```
Comment (cshewmake2) [inline sparsecoding/inference.py:365]: Move comment above line

**Review by cshewmake2 — COMMENTED** [review]
Overall fantastic! Just a few tweaks on naming conventions

File: sparsecoding/inference.py (hunk 1 above)
Comment (denkle) [reply sparsecoding/inference.py]: Modified the names accordingly

File: sparsecoding/inference.py (hunk 4 above)
Comment (denkle) [reply sparsecoding/inference.py]: Used lambdas instead

**Pull request merged**